src/abstractdataframe/reshape.jl

```diff
@@ -399,3 +399,40 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
     res.refs = repeat(res.refs, inner = [v.inner], outer = [v.outer])
     res
 end
+
+
+"""
+    transpose(df::AbstractDataFrame, src_namescol=1, dest_namescol=names(df)[src_namescol]; copycols::Bool=false, makeunique=false, promote_type::Bool=true)
+
+Transpose a `DataFrame`, such that rows become columns,
+and the column indexed by `src_namescol` becomes a header.
+
+By default, the type of resulting columns will depend on the promoted type of all input columns.
+Pass `promote_type=false` to make resulting column types depend on the elements of the originating rows,
+though note that this may be substantially slower.
+
+
+# Examples
+
+```julia
+df1 = DataFrame(a=["x", "y"], b=rand(2), c=[1,2], d=rand(Bool,2)) # all types can be promoted to Float64
+df2 = DataFrame(a=["x", "y"], b=[1, "str"], c=[1,2], d=rand(Bool,2))
+
+transpose(df1)
+transpose(df1, promote_type=false)
+
+transpose(df2)
+transpose(df2, promote_type=false)
+````
+"""
+function Base.transpose(df::AbstractDataFrame, src_namescol=1, dest_namescol=names(df)[src_namescol]; copycols::Bool=false, makeunique=false, promote_type::Bool=true)
+    if promote_type
+        m = permutedims(Matrix(df[!, Not(src_namescol)]))
+        df2 = DataFrame([names(df[!, Not(src_namescol)])], [dest_namescol])
+        return hcat(df2, DataFrame(m, df[!, src_namescol], makeunique=makeunique), copycols=copycols)
```

Comment:
better use `hcat!`, `copycols` can be `false` here. Also note that `hcat!` also has `makeunique` kwarg that should be used.

Also - as commented `DataFrame(m, df[!, src_namescol], makeunique=makeunique)` will soon be deprecated, so we should write something like `df_tmp = rename!(DataFrame(Tables.table(m)), df[!, src_namescol]))` and then use it in `hcat!`.

Finally the decision is if we require `df[!, src_namescol]` to hold `Symbol` or `AbstractString` (I think yes) or we should do a conversion to `Symbol` of whatever is stored in this column.